pkg/util/httputil/loggedclient.go

```diff
@@ -0,0 +1,42 @@
+package httputil
+
+import (
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// NewLoggingClient returns the http.Client with wrapped Transport that is capable of logging request
+func NewLoggingClient(client http.Client, log *zap.SugaredLogger) http.Client {
+	return http.Client{
+		Transport:     &loggedRoundTripper{rt: client.Transport, log: log},
+		CheckRedirect: client.CheckRedirect,
+		Jar:           client.Jar,
+		Timeout:       client.Timeout,
+	}
+}
+
+type loggedRoundTripper struct {
+	rt  http.RoundTripper
+	log *zap.SugaredLogger
+}
```

Comment:
[nit] I think this may be a bit wasteful, `digest` already gave you a client and you're now copying it over to a new client, I would probably embed `digest.Transport` here
like
```go
type LoggingTransport struct {
	digest.Transport
	log *zap.SugaredLogger
}

func NewLoggingTransport(username, password string, log *zap.SugaredLogger) *LoggingTransport {
	return &LoggingTransport{
		Transport:     digest.Transport{
			Username: username,
			Password: password,
			Transport: http.DefaultTransport,
		},
		log: log,
	}
}

func (t *LoggingTransport) RoundTrip(request *http.Request) (*http.Response, error) {
	startTime := time.Now()
	response, err := t.Transport.RoundTrip(request)
	duration := time.Since(startTime)
	t.logResponse(request, response, err, duration)
	return response, err
}

// to be used like

t := httputil.NewLoggingTransport(connection.PublicKey, connection.PrivateKey, log)
tc, err := t.Client()
if err != nil {
	return nil, err
}
mongodbatlas.New(tc, mongodbatlas.SetBaseURL("https://cloud-qa.mongodb.com/api/atlas/v1.0/"))
```

I would even consider making `NewLoggingTransport(username, password string, log *zap.SugaredLogger)` receive options in a functional way but this can de done later. My point here is that it feels like a lot of passing around a client (copy way) when just one should be enough 

(some options in my example could be tuned based on pointers or not, I don't have a strong opinion on the details of my example just the point of reusing transport instead of creating a new client)

Edit: reusing `t.Client()` may not work

Reply:
thanks for the extended review!
I see your point and I think it has good reasons and explanation.
The reasons why I like docorating pattern more here is that this LoggingClient can optionally "wrap" another HTTP client and not couple with other functionality (digest in this case). Initially I even wanted to make "digest client" a decorator over the HTTP client as well but `digest` library is built a bit different way so I decided to make it a "starting point" of creating an HTTP client (this is ok as the default go Http client keeps the global cache etc state that I wasn't aware of). So now the digest client is initialized in the `atlas/client.go` and not in this `util` package which contains only the logging wrapper.
As to copying the structs - I believe it's a very [cheap operation](https://go101.org/article/value-copy-cost.html#:~:text=The%20costs%20of%20copying%20small,are%20all%20small%2Dsize%20types.&text=The%20costs%20of%20copying%20values,base%20types%20are%20very%20small.) and afaik is even cheaper than working with pointers

Reply:
on the subject of pointers vs value, it's cheap for basic types, for advanced structs it depends of the struct as each property adds up to the cost so big structs can fall in the not cheap side. This can also be seeing in your link where they benchmark copy vs pointers, copy is twice as slow as pointers for a struct*
```
Benchmark_Range_OneIterVar-4    500000   3203 ns/op
Benchmark_Range_TwoIterVars-4   200000   6616 ns/op
```
* they mention this and I agree, it will always depend on multiple things

Also understanding the flow of digest.Client to LoggedClient requires some mental exercise here because you are still decorating transport even if you don't want to tightly couple it with `digest.Transport`, `RoundTrip` is for Transport and not client.

I agree with you if you don't want to tightly couple this with `digest` but to avoid this another option can be 

```go
type LoggingTransport struct {
	http.RoundTripper
	log *zap.SugaredLogger
}

func NewLoggingClient(t http.RoundTripper, log *zap.SugaredLogger) *http.Client {
	return &http.Client{
		Transport: &LoggingTransport{RoundTripper: t, log: log},
	}
}

func (t *LoggingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	startTime := time.Now()
	res, err := t.RoundTripper.RoundTrip(req)
	duration := time.Since(startTime)
	t.logResponse(request, response, err, duration)
	return res, err
}
```

Again the point here is to avoid so many http.Client going around (no strong opinion here if you want `RoundTripper` embedded or as a prop)

Reply:
As discussed p2p the functional approach could be an alternative to flexible configuration of the client.